sw/host/opentitanlib/src/rom_patch/patch.rs

```diff
@@ -0,0 +1,745 @@
+                patch_code.push(insn_hex);
+            }
+
+            if let Some(sub_routines_hjson) = &region.sub_routines {
+                for sub_routine_hjson in sub_routines_hjson {
+                    let sub_routine = RomPatchSubRoutine::from_hjson(sub_routine_hjson, base_addr)?;
+                    sub_routines.push(sub_routine);
+                }
+            }
+        }
+
+        let mut code_addr_end = base_addr + ((patch_code.len() as u32) * 4);
+        for s in &sub_routines {
+            code_addr_end = std::cmp::max(code_addr_end, s.range().end);
+        }
+
+        let code_len = (code_addr_end - base_addr) as usize;
+        let patch_size = mem::size_of::<RomPatchHeader>()
+            + (N_MATCH_DESCRIPTORS * mem::size_of::<RomPatchMatchDescriptor>())
+            + code_len
+            + RSA_SIGNATURE_LEN_BYTES;
+
+        let header = RomPatchHeader {
+            meta,
+            size: (patch_size >> 2) as u16,
+            revision: patch.revision,
+        };
+
+        Ok(RomPatch {
+            header,
+            size: patch_size,
+            code_offset: PATCH_CODE_SECTION_OFFSET,
+            signature_offset: PATCH_CODE_SECTION_OFFSET + code_len,
+            base_addr,
+            match_table,
+            patch_code,
+            sub_routines,
+            signature_key,
+        })
+    }
+
+    fn code_offset(&self) -> usize {
+        self.code_offset
+    }
+
+    fn signature_offset(&self) -> usize {
+        self.signature_offset
+    }
+
+    fn code_len(&self) -> usize {
+        self.signature_offset() - self.code_offset()
+    }
+
+    fn len(&self) -> usize {
+        self.size
+    }
+
+    fn validate(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl<'a> fmt::Display for RomPatch<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(
+            f,
+            "Code range [{:#x}:{:#x}]",
+            self.base_addr,
+            self.base_addr + self.code_len() as u32
+        )?;
+        writeln!(f, "Header\n{}", self.header)?;
+        for (i, m) in self.match_table.iter().enumerate() {
+            if !m.enabled() {
+                continue;
+            }
+            writeln!(f, "Match descriptor #{}: \n{}", i, m)?;
+        }
+        for (i, s) in self.sub_routines.iter().enumerate() {
+            writeln!(f, "Sub routine #{}: \n{}", i, s)?;
+        }
+        writeln!(f, "Remapped code {:?} bytes", self.patch_code.len() * 4)?;
+        writeln!(f, "Size {:?} bytes", self.len())?;
+        writeln!(f, "\tHeader {:?} bytes", mem::size_of::<RomPatchHeader>())?;
+        writeln!(
+            f,
+            "\tDescriptors {:?} bytes",
+            mem::size_of_val(&self.match_table)
+        )?;
+        writeln!(f, "\tCode {:?} bytes", self.code_len())?;
+        writeln!(f, "\tSignature {:?}", RSA_SIGNATURE_LEN_BYTES)
+    }
+}
+
+impl<'a> Read for RomPatch<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```

Comment:
As @jwnrt said above, this could use `zerobytes::FromBytes` to make it safer and clearer. See an example here:
https://github.com/lowRISC/opentitan/pull/20672/files#diff-4455f74af62f7ea7ec84c7fd2e934a0b14e940e9d917ae899d276fc5970968afR315

Reply:
Yes. The updated PR is now based on the `zerocopy` crate.